### PR TITLE
Add new vote state version that replaces Lockout with LandedVote to a…

### PR DESCRIPTION
**Pull request by bji**
…llow

vote latency to be tracked in a future change.  Includes a feature to be enabled which will when enabled cause the vote state to be written in the new form.

#### Problem

See timely vote credits proposal.

#### Summary of Changes

Added new VoteStateVersions Current.  Moved existing Current to 1_14_11.  Same for Tower.  Added feature for enabling storage of new vote state version.  Implemented method for updating storage to new vote state version only after the feature is enabled, resizing the vote account as necessary.   Modified tower storage so that its storage format is fixed at the 1_14_11 version so that tower files across software versions are compatible.  Other mechanical changes to support these modifications.

### CODE REVIEW

**Comment by bji** [comment]
The basic idea here is to make room for storage of vote latency.  The field added by this change (LandedVote.latency) will be utilized in a follow on change that implements the timely vote credits feature.  That implementation will basically reproduce the remainder of this change: https://github.com/solana-labs/solana/pull/28126.

This implementation will be able to read in a VoteState of any version and always converts it to the Current version.  However, when writing vote state out (by the vote program itself, when it updates its state), if the vote_state_add_vote_latency feature is enabled, the Current version will be written out (which includes the new LandedVote struct in place of the Lockouts struct in the vote state); if the feature is not enabled, the "old" version of the vote state will be written out.

This means that:

- While validators are upgrading to the new software version, the persisted version of vote state will be unaffected.  This allows snapshots to remain compatible between all software versions.

- After upgrading to the new software version, vote state will always be operated on at runtime in its "new" form (but with the same functionality since the new field is unused), but will be written out in its "old" form.  This is slightly inefficient as there will be an extra step on every vote instruction: deserialize the old form and convert to the new form; and on saving the vote state, the corresponding conversion to the old form before serializing and writing out.  But this mechanism retains compatibility for account state in snapshots.

- After upgrading to the new software, the Tower file format will change for the validator -- it will be updated to the new version after being read in, and will be written out in the new form (with its internal vote state that it keeps for the validator's vote account being of the new form).  Since this file is generally only stored locally, this doesn't matter, except that it means that a downgrade to an older software version will not be able to load the tower file.  This makes upgrading to the new software a one-way street, unless the validator operator is willing to delete their tower file on downgrade.  Since downgrade requires a validator restart, and since that takes so long that validators are very unlikely to still be in lockout after they've begun voting again, blowing away the tower and starting over on downgrade is "probably ok".

- After the feature is enabled, the new vote state will be written on the next vote instruction for each vote account.  This will eliminate the cost of the conversions from old to current and back to old when reading/writing vote account state.  But it will make that validator's snapshots no longer usable by any validator still on the old software version.  For this reason, the feature should definitely only be enabled after a large percentage of validators have already upgraded.

- After the feature is enabled, any vote account which does not have enough rent exempt minimum to allow the vote account to be resized to be 31 bytes larger, will continue to write the old vote state version.  As soon as the rent exempt minimum requirement is met for that vote account, the new state will be written (on the next vote instruction).  The lamports required to fund all vote accounts for this is really trivial; much less than 1 SOL.  I would be willing to just personally fund all the known vote accounts in preparation for flipping on the feature switch.

So to summarize, the tricky issues are:

- The software is not downgrade-able while retaining the local tower file.  Downgrading would require deleting the tower.  However, downgrade is really not expected (how often do validators downgrade anyway?).  And furthermore, deleting the tower on restart is very unlikely to lead to lockout violations given the time it takes to restart.

- There will be a slight performance penalty for data format conversions out of and back to the persisted state, on each vote program instruction, until the feature is enabled.  But the effect is probably quite minor.

- After the feature is enabled, vote accounts must be properly funded before they can upgrade their vote account state.  But this is really easy to achieve with a minimal amount of SOL.

**Comment by bji** [comment]
A note: I could make a tower file down-convert program that could assist if there is a need to downgrade from software containing this change to older software.  That could alleviate any concerns about downgrade-ability, but would require special steps on the part of the validator operator to successfully complete a downgrade.

File: programs/vote/src/vote_state/mod.rs [hunk 1]
```diff
@@ -141,6 +141,37 @@ pub fn to<T: WritableAccount>(versioned: &VoteStateVersions, account: &mut T) ->
...
+    vote_account: &mut BorrowedAccount,
+    vote_state: VoteState,
+    feature_set: &FeatureSet,
+) -> Result<(), InstructionError> {
+    // Only if vote_state_add_vote_latency feature is enabled should the new version of vote state be stored
+    if feature_set.is_active(&feature_set::vote_state_add_vote_latency::id()) {
+        // If the account is not large enough to store the vote state, then attempt a realloc to make it large enough.
+        // The realloc can only proceed if the vote account has balance sufficient for rent exemption at the new size.
+        if (vote_account.get_data().len() < VoteState::size_of())
+            && (!vote_account.is_rent_exempt_at_data_length(VoteState::size_of())
```
Comment (carllin) [inline programs/vote/src/vote_state/mod.rs:156]: In these cases where these accounts have insufficient lamports, they will receive fewer rewards?

So it will be up to the validator operator to check their vote accounts have sufficient balance after this change is released. Probably need to make some comms around this to allow validators to easily check if their accounts have been upgraded to the new version, and catch cases where they could fail:

1. they have sufficient balance
2. The `!vote_account.set_data_length(VoteState::size_of()).is_ok())` check below didn't silently fail (not sure of all the possibilities here)

File: ledger/src/blockstore.rs [hunk 2]
```diff
@@ -4316,49 +4316,49 @@ fn adjust_ulimit_nofile(_enforce_ulimit_nofile: bool) -> Result<()> {
...
 
 #[cfg(unix)]
-fn adjust_ulimit_nofile(enforce_ulimit_nofile: bool) -> Result<()> {
+fn adjust_ulimit_nofile(_enforce_ulimit_nofile: bool) -> Result<()> {
     // Rocks DB likes to have many open files.  The default open file descriptor limit is
     // usually not enough
     // AppendVecs and disk Account Index are also heavy users of mmapped files.
     // This should be kept in sync with published validator instructions.
     // https://docs.solana.com/running-validator/validator-start#increased-memory-mapped-files-limit
-    let desired_nofile = 1_000_000;
```
[the remark this answers is not in the document]
Comment (bji) [reply ledger/src/blockstore.rs:4325]: Yes, thanks for noticing.  I comment out most of that file when I run local tests because I couldn't figure out how to prevent this code from failing while I was running tests.  When I restored the code I must have made a mistake and didn't notice before checking in.  Will fix, thanks for noticing!

File: programs/vote/src/vote_state/mod.rs (hunk 1 above)
Comment (bji) [reply programs/vote/src/vote_state/mod.rs:156]: What will happen is that if the accounts still have insufficient lamports by the time the *second* change is released and its feature enabled, which would actually implement timely vote credits, then yes, those validators will receive a single vote credit per vote, which is the minimum possible.

I believe that this should be easily avoided though because the enablement of the second feature would come a significant time after the enablement of the feature implemented in this current change, giving validators plenty of time to provide sufficient lamports into their vote account to allow the state upgrade.

Comment (bji) [reply programs/vote/src/vote_state/mod.rs:156]: I could log if the `vote_account.set_data_length(VoteState::size_of()` call fails.  That would at least leave some record of the failure easily visible in a place like the solana explorer for validators who want to figure out why they might not be getting updated state.  I didn't do it because I wasn't sure if logging in the vote program would be an issue since its a very frequently run program and I didn't want to increase the storage cost on the explorer for retaining details about the transaction (including the log messages).

Comment (bji) [reply programs/vote/src/vote_state/mod.rs:156]: In addition, once this current change's feature is switched, we can survey vote accounts to see how many haven't upgraded, and only proceed with the timely vote credits feature once all of the (or the vast majority of the) active vote accounts have successfully upgraded to the new state.

File: core/src/tower_storage.rs [hunk 3]
```diff
@@ -418,4 +435,44 @@ pub mod test {
...
+        let tower_path = TempDir::new().unwrap();
+        let identity_keypair = Keypair::new();
+        let node_pubkey = identity_keypair.pubkey();
+        let mut vote_state = VoteState1_14_11::default();
+        vote_state
+            .votes
+            .resize(MAX_LOCKOUT_HISTORY, Lockout::default());
+        vote_state.root_slot = Some(1);
+
+        let vote = Vote::new(vec![1, 2, 3, 4], Hash::default());
```
Comment (AshwinSekar) [inline core/src/tower_storage.rs:450]: nit: make this a `VoteStateUpdate` since that's what will be used at the time of migration

File: sdk/program/src/vote/state/mod.rs [hunk 4]
```diff
@@ -302,6 +327,26 @@ impl VoteState {
         })
     }
 
+    // This helper function exists to serialize the current vote state into a v1_14_11 version, to be used to store
+    // the old form of vote_state until the following are both true for a given vote account: a) the
+    // vote_state_add_vote_latency feature is enabled, and b) the vote account is either already large enough to hold
+    // the current vote state, or the vote account can be successfully re-allocated to be large enough for the current
+    // vote state
+    pub fn as_v1_14_11(&self) -> vote_state_1_14_11::VoteState1_14_11 {
```
Comment (AshwinSekar) [inline sdk/program/src/vote/state/mod.rs:335]: can we make this a From impl?

File: programs/vote/src/vote_state/mod.rs [hunk 5]
```diff
@@ -141,6 +141,37 @@ pub fn to<T: WritableAccount>(versioned: &VoteStateVersions, account: &mut T) ->
     VoteState::serialize(versioned, account.data_as_mut_slice()).ok()
 }
 
+// Updates the vote account state with a new VoteState instance.  This is required temporarily during the
+// upgrade of vote account state from V1_14_11 to Current.
+fn set_vote_account_state(
```
Comment (AshwinSekar) [inline programs/vote/src/vote_state/mod.rs:146]: lets add a test for this to verify the correct format is being stored and properly read back with `vote_account.get_state::<VoteStateVersions>()`

**Comment by mvines** [comment]
> The software is not downgrade-able while retaining the local tower file. Downgrading would require deleting the tower. However, downgrade is really not expected (how often do validators downgrade anyway?). 

Not sure if this represents the current state of the PR but downgradability is a must, without a manual step of deleting the tower. Upgrade/downgrade cycles are common when the beta branch is close to becoming stable.

> And furthermore, deleting the tower on restart is very unlikely to lead to lockout violations given the time it takes to restart.

If we believed this argument then there'd be no need to persist the tower at all.

File: sdk/program/src/vote/state/mod.rs (hunk 4 above)
Comment (bji) [reply sdk/program/src/vote/state/mod.rs:335]: Sure, will do.

File: programs/vote/src/vote_state/mod.rs (hunk 5 above)
Comment (bji) [reply programs/vote/src/vote_state/mod.rs:146]: Yes, I was planning to write tests like that, but was waiting to make sure that this change's overall shape is acceptable before proceeding with that effort.  Since I haven't heard any push back about the overall shape of the change, I'll add the tests.  Thanks!

**Comment by bji** [comment]
> 

How would you recommend proceeding?  How has tower version upgrade been handled in the past?

**Comment by mvines** [comment]
> How would you recommend proceeding? How has tower version upgrade been handled in the past?

Yep and there's breadcrumbs in the code about how a new version could be added. Check out `tower_storage.rs`, maybe starting around here: https://github.com/solana-labs/solana/blob/9cd3bb0c4de929d636ba6350141ee948d829fd12/core/src/tower_storage.rs#L18-L21

> Would writing the old tower version until the feature is flipped be acceptable?

That seems best, although not sure if FeatureSet is plumbed fully to make this super easy.
But this way one can upgrade/downgrade between releases cleanly, until the feature is activated and downgrading is no longer possible.

> Also, is a tool that would downgrade the tower file an acceptable solution, or is this something that has to exist in the validator code itself?

Ideally in the validator itself, so operators don't need to do anything more than what they already do (just swap `solana-validator` binaries)

**Comment by bji** [comment]
> That seems best, although not sure if FeatureSet is plumbed fully to make this super easy. But this way one can upgrade/downgrade between releases cleanly, until the feature is activated and downgrading is no longer possible.


Yes, it's the plumbing of FeatureSet that I a) was hoping to avoid since it's work, and b) since it wasn't already done, and yet there has been a new Tower version at least once, would indicate that the feature set mechanism was not used the last time the tower version was bumped.

However, I'll plumb the feature set stuff as the solution this time.

**Comment by AshwinSekar** [comment]
Think you already have the rooted bank in voting_service, should be straightforward to grab the feature from there and pass it to tower storage

**Comment by bji** [comment]
> Think you already have the rooted bank in voting_service, should be straightforward to grab the feature from there and pass it to tower storage. Last time i used different filenames for tower version and ran the conversion if the file wasn't there.

This newest version doesn't change the format of the saved tower, so it bypasses the concerns.

**Comment by bji** [comment]
> > The software is not downgrade-able while retaining the local tower file. Downgrading would require deleting the tower. However, downgrade is really not expected (how often do validators downgrade anyway?).
> 
> Not sure if this represents the current state of the PR but downgradability is a must, without a manual step of deleting the tower. Upgrade/downgrade cycles are common when the beta branch is close to becoming stable.

This newest version doesn't change the format of the saved tower, so it bypasses the concerns.

File: sdk/program/src/vote/state/mod.rs (hunk 4 above)
Comment (bji) [reply sdk/program/src/vote/state/mod.rs]: This is done in the newest version.

File: sdk/program/src/vote/state/mod.rs [hunk 6]
```diff
@@ -88,6 +89,30 @@ impl Lockout {
...
+
+impl From<Lockout> for LandedVote {
+    fn from(lockout: Lockout) -> Self {
+        Self {
+            latency: 0,
+            lockout,
+        }
+    }
+}
+
```
Comment (t-nelson) [inline sdk/program/src/vote/state/mod.rs:115]:
```suggestion

impl LandedVote {
    pub fn slot(&self) -> Slot {
        self.lockout.slot
    }
    
    pub fn confirmation_count(&self) -> u32 {
        self.lockout.confirmation_count
    }
}
```

see https://github.com/solana-labs/solana/pull/29753.  now all of the noisy plumbing can go away

Comment (bji) [reply sdk/program/src/vote/state/mod.rs:115]: This is fine; although I personally think that automatic conversions like this are on balance not  beneficial, as they don't mean anything in isolation and thus defeat readability in places where I think being explicit leads to better self-documentation of code.  But I'm definitely not going to argue with anyone at Solana Labs about this stuff, I'm just your lackey at this point.

Comment (t-nelson) [reply sdk/program/src/vote/state/mod.rs:115]: it's a code maintenance advantage.  going to get rid of a bunch of noise here and ease review

File: core/src/tower_storage.rs [hunk 7]
```diff
@@ -94,7 +109,10 @@ impl SavedTower {
             )));
         }
 
-        let data = bincode::serialize(tower)?;
+        // SavedTower always stores its data in 1_14_11 format
```
Comment (AshwinSekar) [inline core/src/tower_storage.rs:112]: Is the idea here that the latency bit is only needed on chain so it's fine to just stick to the original tower format locally?

Comment (bji) [reply core/src/tower_storage.rs:112]: Yes, the tower doesn't care about vote credits and it doesn't need to track latency.  In my opinion having the tower use the vote account's data structures is problematic as it conflates the use of those data structures between two very different subsystems.  It makes the tower dependent on the vote program, which is awkward.  The change I have made divorces the on-disk data format of the tower from that of the vote program, which I think is a step forward in addressing this.

Comment (AshwinSekar) [reply core/src/tower_storage.rs:112]: yeah I agree, tower barely uses anything in vote_state apart from the Vec<Lockout>, will be a pita but would be nice to separate.

Comment (bji) [reply core/src/tower_storage.rs:112]: I have added a test in the newest version of this change, to verify that the expected vote account state is written under various circumstances.

**Comment by bji** [comment]
Would it be possible to at least get the CI label added so that CI can run on this change?

**Comment by t-nelson** [comment]
the `CI` label only works for a single run.  there's also `ci/run-local.sh`, which you can run... locally, to the same effect (mostly).

**Comment by bji** [comment]
> the `CI` label only works for a single run. there's also `ci/run-local.sh`, which you can run... locally, to the same effect (mostly).

Yeah I tried running that but it wants to fetch master from origin and I really didn't feel like figuring out how to either use it or hack it so that it runs my stuff instead of master, but I guess I will make the effort now.

**Comment by t-nelson** [comment]
> > the `CI` label only works for a single run. there's also `ci/run-local.sh`, which you can run... locally, to the same effect (mostly).
> 
> Yeah I tried running that but it wants to fetch master from origin and I really didn't feel like figuring out how to either use it or hack it so that it runs my stuff instead of master, but I guess I will make the effort now.
> 
> EDIT: OK wasn't so hard. Just some stuff to rip out of test-sanity.sh ...

AFAIK it's only fetching the tip of the base branch so it can check for conflict markers. the actual tests are run against the local checkout

i think that check could be improved to not make so many assumptions about the local configuration though...

**Comment by bji** [comment]
The local_cluster_slow_1 test succeeds locally.  I believe it may be an unreliable test that fails due to timing conditions.

**Comment by bji** [comment]
And now the local_cluster_flakey test is failing despite working locally.  I think the CI system is flaky.

**Comment by t-nelson** [comment]
> And now the local_cluster_flakey test is failing despite working locally. I think the CI system is flaky.

yep, that's why we named it that

File: core/src/consensus.rs [hunk 8]
```diff
@@ -139,6 +141,7 @@ pub(crate) struct ComputedBankState {
 #[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
 pub enum TowerVersions {
     V1_17_14(Tower1_7_14),
+    V1_14_11(Tower1_14_11),
```
Comment (t-nelson) [inline core/src/consensus.rs:144]: are the tower changes strictly required to reserve the new space in `VoteState`?  i'm not seeing any dependencies

we should defer them to the changes that implement the actual timely vote credits feature if so

File: core/src/consensus.rs [hunk 9]
```diff
@@ -986,24 +1013,24 @@ impl Tower {
     ) -> bool {
         let mut vote_state = self.vote_state.clone();
         process_slot_vote_unchecked(&mut vote_state, slot);
-        let vote = vote_state.nth_recent_vote(self.threshold_depth);
-        if let Some(vote) = vote {
-            if let Some(fork_stake) = voted_stakes.get(&vote.slot()) {
-                let lockout = *fork_stake as f64 / total_stake as f64;
+        let lockout = vote_state.nth_recent_lockout(self.threshold_depth);
```
Comment (t-nelson) [inline core/src/consensus.rs:1016]: try to avoid making cosmetic changes like this rename in the same PR as logical changes.  they make review more difficult from both sides

File: core/src/consensus.rs [hunk 10]
```diff
@@ -3161,8 +3207,8 @@ pub mod test {
     #[test]
     fn test_adjust_lockouts_after_replay_time_warped() {
         let mut tower = Tower::new_for_tests(10, 0.9);
-        tower.vote_state.votes.push_back(Lockout::new(1));
-        tower.vote_state.votes.push_back(Lockout::new(0));
+        tower.vote_state.votes.push_back(Lockout::new(1).into());
```
Comment (t-nelson) [inline core/src/consensus.rs:3210]: generally prefer `IntoType::from(from_value)` over `from_value.into()` for the improved self-documentation

**Review by t-nelson — COMMENTED** [review]
vote state stuff is looking good. prefer to defer the tower changes if they aren't strictly required here (as commented below)

File: core/src/consensus.rs (hunk 8 above)
Comment (bji) [reply core/src/consensus.rs:144]: The issue is that if the tower modifications here are not made, then the serialized format of the tower changes when the VoteState struct is changed, since the tower stores a VoteState instance.  This then means that upgrading to the version of validator software that includes this change will start writing a tower file that on downgrade will be unreadable by the downgraded-to software.

Therefore I have altered the tower logic to continue to serialize the tower using the older serialized format.  I don't see any way around this if the goal of ensuring that downgrade is not broken is to be achieved.

File: core/src/consensus.rs (hunk 9 above)
Comment (bji) [reply core/src/consensus.rs:1016]: I understand the point but this is a minor change and pretty easy to follow if you try.  Additionally, the lack of consistency in naming makes the code altogether harder to read.  The change of name to nth_recent_lockout is meant to ensure that the logical purpose of the code change in this commit can be followed.

File: core/src/consensus.rs (hunk 10 above)
Comment (bji) [reply core/src/consensus.rs:3210]: That is a good idea.  I'll make the change.

File: core/src/consensus.rs (hunk 9 above)
Comment (bji) [reply core/src/consensus.rs:1016]: By the way if there is any chance of this being a blocking issue, please let me know and I'll move that rename into a separate change.

**Comment by AshwinSekar** [comment]
VoteState changes and the upgrade test looks good to me. I'll ping carl to see if he has any concerns, but hopefully we can ship soon.

File: core/src/consensus.rs (hunk 8 above)
Comment (bji) [reply core/src/consensus.rs:144]: I guess an alternative would be to write the new tower file into a new location, similar to what was done with the tower-1.9 file.  Both versions of the tower (tower-1.9 and tower-1.14) would be written at the same time, until the old file is deprecated later.

However, this would still be changes to the tower, so it doesn't avoid code modifications.  In addition, it's a more complex change that is unnecessary because there is no new data needed in the tower file.  If the tower actually had new data then this would be the necessary approach.

So I was inaccurate when I said "I don't see any way around this", I should have said "I don't see a better way around this".

File: programs/vote/src/vote_state/mod.rs [hunk 11]
```diff
@@ -1114,6 +1156,98 @@ mod tests {
         )
     }
 
+    #[test]
+    fn test_vote_state_upgrade_from_1_14_11() {
```
Comment (carllin) [inline programs/vote/src/vote_state/mod.rs:1160]: Thanks for this test!

File: programs/vote/src/vote_state/mod.rs [hunk 12]
```diff
@@ -1114,6 +1156,98 @@ mod tests {
...
+        assert_eq!(
+            borrowed_account.set_lamports(rent.minimum_balance(VoteState::size_of())),
+            Ok(())
+        );
+        assert_eq!(
+            set_vote_account_state(&mut borrowed_account, vote_state, &feature_set),
+            Ok(())
+        );
+        let vote_state_version = borrowed_account.get_state::<VoteStateVersions>().unwrap();
+        assert!(matches!(vote_state_version, VoteStateVersions::Current(_)));
```
Comment (carllin) [inline programs/vote/src/vote_state/mod.rs:1248]: Might be good to verify here that the fields are accurate and reflect the old state as well. For that purpose, it would be good to have old state not have default fields

**Review by carllin — COMMENTED** [review]
Just a few nits, looking good!

File: programs/vote/src/vote_state/mod.rs (hunk 12 above)
Comment (bji) [reply programs/vote/src/vote_state/mod.rs:1248]: I will do this, but am traveling this upcoming week so this will be delayed until after that, thanks!

Comment (bji) [reply programs/vote/src/vote_state/mod.rs:1282]: Hi sorry for the delay, I have updated the test to include nonzero vote state and verify that the vote state has not changed throughout the upgrade tests.

File: core/src/consensus.rs (hunk 8 above)
Comment (bji) [reply core/src/consensus.rs:144]: Yeah, gonna be especially problematic when the 1.17 branch actually exists.

**Review by carllin — DISMISSED** [review]

File: programs/vote/src/vote_state/mod.rs [hunk 13]
```diff
@@ -1114,6 +1156,138 @@ mod tests {
...
+        assert!(matches!(vote_state_version, VoteStateVersions::V1_14_11(_)));
+
+        // Convert the vote state to current as would occur during vote instructions
+        let converted_vote_state = vote_state_version.convert_to_current();
+
+        // Check to make sure that the vote_state is unchanged
+        assert!(vote_state == converted_vote_state);
+
+        let vote_state = converted_vote_state;
+
```
[the remark this answers is not in the document]
Comment (bji) [reply programs/vote/src/vote_state/mod.rs:1235]: Can we make that a follow-up change?  To be honest I'd make strong arguments that it's not more readable to break it up, but I'd rather not have that debate when this change is presumably *so close* ...

[the remark this answers is not in the document]
Comment (carllin) [reply programs/vote/src/vote_state/mod.rs:1235]: there's a lot of 

```
assert!(vote_state == converted_vote_state);
 let vote_state = converted_vote_state;
```

which is a bit confusing to read and could be simplified if the test cases were self contained.

But yes, a separate PR is fine 😄

**Review by AshwinSekar — APPROVED** [review]
Just changed the pubkey to one I own

**Review by carllin — APPROVED** [review]

**Comment by Lichtso** [comment]
This requires a feature gate issue, which also should be referenced in sdk/src/feature_set.rs
